### ISSUE TITLE
Wrong result if variable and function have the same name

### ISSUE THREAD

**Issue by dktcoding** [fasseg/exp4j#29]
I've been investigating how to improve the performance of variable value retrieval, and came across something odd, the variable names are checked in ```Expression#setVariable()```, but they are not checked before building the expression, that is, ```ExpressionBuilder``` passes the variable names to ```ShuntingYard``` and then they are passed to ```Tokenizer``` without any kind of check. Since ```Tokenizer``` checks if a token is a variable before it checks the function names the process seems to work, ```Expression.evaluate()``` even yields a result... **BUT** the value is wrong.

As an example, the following code should throw a ```IllegalArgumentException``` but it fails silently.

```java
    public void test() {
        Expression e = new ExpressionBuilder("log10(log10)")
                .variables("log10")
                .build();
        e.setVariable("log10", 1);
        System.out.println(e.evaluate()); //The output should be 0 but it's 1
    }
```

I'll send a patch, but the only way I see to fix this issue involves moving around too much code... Basically, moving the validation of variables to ```ExpressionBuilder#build()```, and changing the way that variables are validated and set in ```Expression#setVariable()``` and ```Expression#validate()```, plus moving the values to ```VariableToken``` instead of using a different data structure.

**Comment by fasseg** [comment]
Good catch!

**Comment by fasseg** [comment]
Wouldn't it be easier to check if a variable has the same name as a function during the build process?
We could just iterate over the vars and throw an exception if it's got the same name as a function.

**Comment by dktcoding** [comment]
My bad, that patch does two things:

1. Move the validation of names to ```ExpressionBuilder#build()```
2. Avoid the use of ```HashMap``` to keep the variable values (which is the big performance bottleneck in ```Expression#evaluate()``` now that we use ```ArrayStack```)
3. I forgot to push a patch that creates only one instance of ```VariableToken``` for each variable in ```Tokenizer```

**Comment by fasseg** [comment]
Shouldn't this small patch be enough:
https://github.com/fasseg/exp4j/compare/varfunccheck
It just checks if the there's a function by the same name as a function.

**Comment by dktcoding** [comment]
Yeahp, as I stated in https://github.com/fasseg/exp4j/issues/29#issuecomment-71727271 the idea was also to avoid using ```HashMap``` for storing variables, since using a single object (```VariableToken```) for each variable will avoid a lot of calls to ```HashMap#get()```.

I was short on time and it seemed easier at that moment to push two changes on the same commit (in retrospect it's obvious that they are *completely different* changes, apparently it looked to me as a good idea at the time... Oops...).

**Comment by fasseg** [comment]
Ok so ill have a look again soon and will merge the varfunccheck branch.

**Comment by fasseg** [comment]
Fixed in release 0.4.5
